Update "oc get all" to show resource kinds in output

juanvallejo opened the pull request:
Fixes #2354 
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1247762
UPSTREAM: https://github.com/kubernetes/kubernetes/pull/27379

##### Output

`$ oc get all --all-namespaces`
```
NAMESPACE   NAME              DOCKER REPO   TAGS               UPDATED
default     is/node                         latest             19 hours ago
default     is/ruby                         latest             19 hours ago
default     is/test
NAMESPACE   NAME              REVISION      DESIRED            CURRENT   TRIGGERED BY
default     dc/node           1             1                  1         config,image(node:latest)
default     dc/ruby           1             1                  1         config,image(ruby:latest)
NAMESPACE   NAME              DESIRED       CURRENT            AGE
default     rc/node-1         1             1                  19h
default     rc/ruby-1         1             1                  19h
NAMESPACE   NAME              CLUSTER-IP    EXTERNAL-IP        PORT(S)                 AGE
default     svc/kubernetes    172.30.0.1    <none>             443/TCP,53/UDP,53/TCP   1d
NAMESPACE   NAME              READY         STATUS             RESTARTS                AGE
default     po/node-1-h2vc0   0/1           CrashLoopBackOff   50                      19h
default     po/ruby-1-lsqt6   0/1           CrashLoopBackOff   31                      19h
```

juanvallejo commented:
[test]

fabianofranz commented:
Also fixes https://github.com/openshift/origin/issues/2354.

juanvallejo commented:
@fabianofranz or @kargakis, would you mind taking a look?

juanvallejo commented:
@fabianofranz @kargakis Updated `oc get all` output to display resource type as part of resource name:

```
NAMESPACE   NAME                             TYPE                                          FROM       LATEST
test2       buildconfigs/ruby-sample-build   Source                                        Git        1
NAMESPACE   NAME                             TYPE                                          FROM       STATUS           STARTED      DURATION
test2       builds/ruby-sample-build-1       Source                                        Git        Failed (Error)   9 days ago   39s
NAMESPACE   NAME                             DOCKER REPO                                   TAGS       UPDATED
test2       origin-ruby-sample               172.30.83.241:5000/test2/origin-ruby-sample
test2       ruby-22-centos7                  centos/ruby-22-centos7                        latest     9 days ago
test20      node                             172.30.83.241:5000/test20/node                latest     2 days ago
NAMESPACE   NAME                             REVISION                                      REPLICAS   TRIGGERED BY
default     dc/docker-registry               1                                             1          config
test2       dc/database                      1                                             1          config
test2       dc/frontend                      0                                             2          config,image(origin-ruby-sample:latest)
test20      dc/node                          1                                             1          config,image(node:latest)
NAMESPACE   NAME                             DESIRED                                       CURRENT    AGE
default     rc/docker-registry-1             1                                             1          13d
test2       rc/database-1                    1                                             1          9d
test20      rc/node-1                        0                                             0          2d
test5       rc/test-deployment-config-3      0                                             0          5d
NAMESPACE   NAME                             HOST/PORT                                     PATH       SERVICE                 TERMINATION   LABELS
test2       routes/route-edge                www.example.com                                          frontend                edge          app=ruby-sample-build,template=application-template-stibuildNAMESPACE   NAME      CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
default     svc/docker-registry              172.30.83.241                                 <none>     5000/TCP                13d
default     svc/kubernetes                   172.30.0.1                                    <none>     443/TCP,53/UDP,53/TCP   13d
test2       svc/database                     172.30.130.169                                <none>     5434/TCP                9d
test2       svc/frontend                     172.30.197.30                                 <none>     5432/TCP                9d
NAMESPACE   NAME                             READY                                         STATUS     RESTARTS                AGE
default     po/docker-registry-1-91k21       1/1                                           Running    2                       9h
test2       po/database-1-zh5ad              1/1                                           Running    0                       9h
```

kargakis commented:
Why doesn't it output the resource type for image streams? Also why buildconfigs is not bc?

juanvallejo commented:
[test]

juanvallejo commented:
re[test]

juanvallejo commented:
> Why doesn't it output the resource type for image streams? Also why buildconfigs is not bc?
@kargakis Thanks for catching that, output should now be:

```
_output/local/bin/linux/amd64/kubectl get all --all-namespaces
NAMESPACE   NAME                          DESIRED          CURRENT       AGE
default     rc/docker-registry-1          1                1             14d
test2       rc/database-1                 1                1             9d
test20      rc/node-1                     0                0             2d
test5       rc/test-deployment-config-3   0                0             6d
NAMESPACE   NAME                          CLUSTER-IP       EXTERNAL-IP   PORT(S)                 AGE
default     svc/docker-registry           172.30.83.241    <none>        5000/TCP                14d
default     svc/kubernetes                172.30.0.1       <none>        443/TCP,53/UDP,53/TCP   14d
test2       svc/database                  172.30.130.169   <none>        5434/TCP                9d
test2       svc/frontend                  172.30.197.30    <none>        5432/TCP                9d
NAMESPACE   NAME                          READY            STATUS        RESTARTS                AGE
default     po/docker-registry-1-7xlgg    1/1              Running       0                       7h
test2       po/database-1-f88ba           1/1              Running       4                       7h
NAMESPACE   NAME                          STATUS           VOLUME        CAPACITY                ACCESSMODES   AGE
test5       pvc/pvc-aipl7                 Pending                                                              6d
test5       pvc/pvc-vrj2n                 Pending                                                              6d
```

juanvallejo commented:
> Why doesn't it output the resource type for image streams? Also why buildconfigs is not bc?

@kargakis Thanks for catching that, the output should now be:

```
oc get all --all-namespaces
NAMESPACE   NAME                              TYPE                                          FROM          LATEST
test2       bc/ruby-sample-build              Source                                        Git           1
NAMESPACE   NAME                              TYPE                                          FROM          STATUS           STARTED      DURATION
test2       builds/ruby-sample-build-1        Source                                        Git           Failed (Error)   9 days ago   39s
NAMESPACE   NAME                              DOCKER REPO                                   TAGS          UPDATED
test2       imagestreams/origin-ruby-sample   172.30.83.241:5000/test2/origin-ruby-sample
test2       imagestreams/ruby-22-centos7      centos/ruby-22-centos7                        latest        9 days ago
test20      imagestreams/node                 172.30.83.241:5000/test20/node                latest        2 days ago
NAMESPACE   NAME                              REVISION                                      DESIRED       CURRENT   TRIGGERED BY
default     dc/docker-registry                1                                             1             0         config
test2       dc/database                       1                                             1             0         config
test2       dc/frontend                       0                                             2             0         config,image(origin-ruby-sample:latest)
test20      dc/node                           1                                             1             0         config,image(node:latest)
NAMESPACE   NAME                              DESIRED                                       CURRENT       AGE
default     rc/docker-registry-1              1                                             1             14d
test2       rc/database-1                     1                                             1             9d
test20      rc/node-1                         0                                             0             2d
test5       rc/test-deployment-config-3       0                                             0             6d
NAMESPACE   NAME                              HOST/PORT                                     PATH          SERVICE                 TERMINATION   LABELS
test2       routes/route-edge                 www.example.com                                             frontend                edge          app=ruby-sample-build,template=application-template-stibuild
NAMESPACE   NAME                              CLUSTER-IP                                    EXTERNAL-IP   PORT(S)                 AGE
default     svc/docker-registry               172.30.83.241                                 <none>        5000/TCP                14d
default     svc/kubernetes                    172.30.0.1                                    <none>        443/TCP,53/UDP,53/TCP   14d
test2       svc/database                      172.30.130.169                                <none>        5434/TCP                9d
test2       svc/frontend                      172.30.197.30                                 <none>        5432/TCP                9d
NAMESPACE   NAME                              READY                                         STATUS        RESTARTS                AGE
default     po/docker-registry-1-7xlgg        1/1                                           Running       0                       7h
test2       po/database-1-f88ba               1/1                                           Running       4                       7h
```

juanvallejo commented:
[test]

kargakis commented:
@juanvallejo imagestreams have their own abbreviation too (is)

kargakis commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/kubectl.go:
```diff
@@ -144,7 +144,9 @@ var shortForms = map[string]string{
 	// and add an entry to valid_resources when appropriate.
+	"bc":     "buildconfigs",
 	"cs":     "componentstatuses",
+	"dc":     "deploymentconfigs",
```
These shouldn't be here. We have our own map of abbreviations
https://github.com/openshift/origin/blob/d409b31fa672e2e77940012bca8793a80b838422/pkg/cmd/util/clientcmd/shortcut_restmapper.go#L102

rhcarvalho commented:
I'm happy to see this tackling that old issue. I have a concern with the implementation though.

Wouldn't it be possible to implement it more generically instead of modifying every `print***` function?
The way it is, it's easy to miss some type (as happened, per comments above), and I'd say harder to maintain over time.

juanvallejo commented:
@rhcarvalho 
> Wouldn't it be possible to implement it more generically instead of modifying every print*** function?

Thanks for the feedback! I don't think this would be possible, without perhaps doing type switching to cast and update each generic runtime object before passing it as an argumnt to its `print***` function. This would most likely be done around this part of the code:  https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go#L1844

juanvallejo commented:
@kargakis @fabianofranz 
> These shouldn't be here. We have our own map of abbreviations

Thanks! `pkg/cmd/cli/describe/printer.go` will now use our own abbreviations https://github.com/openshift/origin/pull/9334/commits/b5932e2583e4338bfe2c375c6e7bef80e5fbf279

juanvallejo commented:
[test]

juanvallejo commented:
re[test]

fabianofranz commented:
Yay upstream merged, @juanvallejo make sure this is up-to-date, squash if needed and let's get this in.

juanvallejo commented:
[test]

fabianofranz commented:
[merge]

rhcarvalho commented on pkg/cmd/util/clientcmd/factory.go:
```diff
@@ -205,7 +205,7 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 			return defaultMapper, defaultTyper
 		}
-
+		////--
```
What's this?

juanvallejo replied:
Ah forgot to take this out, it was a quick way of marking where I left off in that file.

rhcarvalho commented on pkg/cmd/util/clientcmd/shortcut_restmapper.go:
```diff
@@ -121,3 +121,7 @@ func expandResourceShortcut(resource unversioned.GroupVersionResource) unversion
 	return resource
 }
+
+func GetResourceShortForms() map[string]string {
```
This is not used anywhere outside of this package, why do we need to export this otherwise unexported `shortForms` map?

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -138,6 +140,11 @@ func NewHumanReadablePrinter(noHeaders, withNamespace, wide bool, showAll bool,
 }
 
+func NewHumanReadablePrinterWithResourceAliases(resourceAliases map[string]string, noHeaders, withNamespace, wide bool, showAll bool, showLabels bool, absoluteTimestamps bool, columnLabels []string) *kctl.HumanReadablePrinter {
+	resourceShortForms = resourceAliases
```
This should be a field of the printer, not mutate the global state (that global variable `resourceShortForms` above).

Calling the function `resourceShortFormFor` before a call to this `NewHumanReadablePrinterWithResourceAliases` will range over an empty map, while after the call it will range over whatever was the last thing passed in as argument.

And if these functions would be called from different goroutines (I believe they will be not, in the CLI code base) there's a data race.

juanvallejo replied:
> This should be a field of the printer, not mutate the global state (that global variable resourceShortForms above).

There is no straightforward way of doing this without changing the function signature of `NewHumanReadablePrinter` (which would involve changing several other files).

> Calling the function resourceShortFormFor before a call to this NewHumanReadablePrinterWithResourceAliases will range over an empty map, while after the call it will range over whatever was the last thing passed in as argument.

It was the most feasible way I found of doing this, short of re-defining the `shotForms` map in the `shortcut_restmapper.go` file. I cannot import the `clientcmd` package here as it causes a circular dependency.

juanvallejo replied:
@rhcarvalho Taking an idea from your suggestions [here](https://github.com/openshift/origin/pull/9334#discussion_r69468339), maybe a clean way to address this would be to edit `pkg/cmd/util/clientcmd/factory.go`, so that instead of [line 379](https://github.com/openshift/origin/pull/9334/files#diff-0073015bc0e88b3269344b5a87d471dbR379) returning the `HumanReadablePrinter` right away, we store it in a `printer` var, and do something similar to:

```go
	// ...
	if mustPrintWithKinds(...) { // mustPrintWithKinds replaces yet another for-loop bloating the over-bloated RunGet.
		printer.EnsurePrintWithKind()  // this method doesn't exist yet, but would set PrintOptions.WithKind to true.
	}
```

juanvallejo replied:
@liggitt Sounds good, I like the idea of passing a single opts struct

liggitt replied:
we even have the opts struct already: `kctl.PrintOptions`

juanvallejo replied:
@smarterclayton Should a PR for this change wait to be merged after the next rebase?

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -159,6 +166,30 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
+		if val == resource {
+			alias = k
+			exists = true
+		}
```
We should break the loop on the first match, shouldn't we?

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -159,6 +166,30 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
 }
 
+// Look-up for resource short forms by value
+func resourceShortFormFor(resource string) (string, bool) {
```
The godoc convention is to start a function documentation with its name:

```go
// resourceShortFormFor looks up for a short form of resource names.
func resourceShortFormFor(resource string) (string, bool) {
```

https://golang.org/doc/effective_go.html#commentary

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/kubectl.go:
```diff
@@ -165,6 +165,19 @@ var shortForms = map[string]string{
 	"svc":    "services",
 }
 
+// Look-up for resource short forms by value
```
@juanvallejo, please take a look at this document it has a lot of useful tips for writing idiomatic Go code :)
https://golang.org/doc/effective_go.html.

rhcarvalho replied:
For documenting functions, the commentary starts with the function name, and goes on as a normal sentence in English, focused on explaining to the user of the function what it is for and how to use it, etc, and end in a full stop.

That's how we get awesome generated documentation like you see in https://docs.golang.org, or even https://godoc.org/github.com/kubernetes/kubernetes (not so great).

Example:

```go
// ResourceShortFormFor returns the short form of a resource name and whether it
// exists. It returns the empty string if there is no short form.
func ResourceShortFormFor(resource string) (string, bool) {
```

Writing the documentation above made me think about using that function, and that maybe it's signature and contract is not as useful as it could be.

Every time you use it, you need to check if the last argument was true, and only then replace some variable with the alias...

What we have now:

```go
			if alias, ok := kubectl.ResourceShortFormFor(mapping.Resource); ok {
				resourceName = alias
			} else if resourceName == "" {
				resourceName = "none"
 			}
```

Alternative:

```go
			resourceName = kubectl.ShortenResourceName(resourceName)
			if resourceName == "" {
				resourceName = "none"
 			}
// ...

// ShortenResourceName returns the short form of a resource name if there is
// any, otherwise returns the name unchanged.
func ShortenResourceName(name string) string {
	for short, resource := range shortForms {
		if name == resource {
			return short
		}
	}
	return name
}
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -333,17 +335,19 @@ type PrintOptions struct {
 type HumanReadablePrinter struct {
 	handlerMap map[reflect.Type]*handlerEntry
-	options    PrintOptions
+	Options    PrintOptions
```
None of the fields of `HumanReadablePrinter` are exported. I think exporting the `Options` and mutating it from outside is a step backwards in the design, it breaks the encapsulation.

rhcarvalho replied:
See how `NewHumanReadablePrinter` takes all other values as arguments. Perhaps changing the function signature to take the other new options explicitly would be better.
@juanvallejo wouldn't that work?

liggitt replied:
Agree on not exporting and definitely not mutating

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -289,9 +291,29 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+		lastMapping = mapping
+	}
+
+	lastMapping = nil
```
This is calling out for a refactor / move it to another function instead of recycling variables.

The following `for` loop below could easily become a function, with a name explaining what it does, and it would return a `[]error`, such that the last lines would read something like:

```go
	// ...
	if mustPrintWithKinds(...) { // mustPrintWithKinds replaces yet another for-loop bloating the over-bloated RunGet.
		printer.EnsurePrintWithKind()  // this method doesn't exist yet, but would set PrintOptions.WithKind to true.
	}
	return utilerrors.NewAggregate(printObjects(printer, ...)) // printObjects replaces the huge final for-loop
}
```

juanvallejo replied:
@rhcarvalho Thanks for the feedback! I will implement this in the next update to this PR

rhcarvalho commented on pkg/cmd/admin/node/node_options.go:
```diff
@@ -33,7 +33,7 @@ type NodeOptions struct {
 	Typer             runtime.ObjectTyper
 	RESTClientFactory func(mapping *meta.RESTMapping) (resource.RESTClient, error)
-	Printer           func(mapping *meta.RESTMapping, noHeaders, withNamespace, wide bool, showAll bool, showLabels bool, absoluteTimestamps bool, columnLabels []string) (kubectl.ResourcePrinter, error)
+	Printer           func(mapping *meta.RESTMapping, printOptions *kubectl.PrintOptions) (kubectl.ResourcePrinter, error)
```
Nice!

rhcarvalho commented on pkg/cmd/admin/node/node_options.go:
```diff
@@ -179,11 +179,31 @@ func (n *NodeOptions) GetPrinters(gvk unversioned.GroupVersionKind) (kubectl.Res
 	}
 
-	printerWithHeaders, err := n.Printer(mapping, false, false, false, false, false, false, []string{})
+	printerWithHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{
```
@juanvallejo `false`, `""`, etc, are the *zero values* of their respective types. Normally we embrace using the zero value, without the need to explicitly initialize it, what would mean that there's no need to for the next several lines.

In code:

```go
// I'd expect Printer to understand a nil printOptions as "use the defaults".
printerWithHeaders, err := n.Printer(mapping, nil)

// Or, if for whatever reason we need to provide a non-nil value:
printerWithHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{})

// Yet, if we want to set just some of the options (just an example):
printerWithHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{Wide: true})
```

rhcarvalho commented on pkg/cmd/admin/node/node_options.go:
```diff
@@ -179,11 +179,31 @@ func (n *NodeOptions) GetPrinters(gvk unversioned.GroupVersionKind) (kubectl.Res
 		return nil, nil, err
 	}
-	printerNoHeaders, err := n.Printer(mapping, true, false, false, false, false, false, []string{})
+	printerNoHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{
```
Here it would be:

```go
printerWithHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{NoHeaders: true})
```

Refactoring the signature of that method eventually is simplifying some code that needed to pass many arguments, now just need to set the options that we need to differ from default :)

rhcarvalho commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -1150,7 +1150,19 @@ func DescribePolicyRule(out *tabwriter.Writer, rule authorizationapi.PolicyRule,
+			AbsoluteTimestamps: false,
+			ColumnLabels:       []string{},
+		}
+		printer := NewHumanReadablePrinter(&printOptions)
```
```go
printerWithHeaders, err := n.Printer(mapping, &kubectl.PrintOptions{NoHeaders: true})
```

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -64,10 +64,12 @@ var (
 	clusterResourceQuotaColumns = []string{"NAME", "SELECTOR"}
 )
 
+var resourceShortForms map[string]string
```
This is not referenced by anything any longer.

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -64,10 +64,12 @@ var (
+func NewHumanReadablePrinter(printOptions *kctl.PrintOptions) *kctl.HumanReadablePrinter {
 	// TODO: support cross namespace listing
-	p := kctl.NewHumanReadablePrinter(noHeaders, withNamespace, wide, showAll, showLabels, absoluteTimestamps, columnLabels)
+	p := kctl.NewHumanReadablePrinter(printOptions)
```
This feels great!

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -159,6 +161,13 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
 }
 
+// Receives a resource kind and its name, returns a formatted string with the
+// "resource_kind_alias/name", or simply "resource_name" if no alias is found
```
This documentation seems to be outdated.

juanvallejo replied:
Ah, I did forget to update this comment, I am now handling our short names here: https://github.com/juanvallejo/origin/blob/1ff8ec94d186fe3770c06cdece2ed8d5d3d9ef99/pkg/cmd/util/clientcmd/factory.go#L379-384

rhcarvalho commented on pkg/cmd/cli/describe/printer_test.go:
```diff
@@ -61,7 +61,18 @@ var MissingPrinterCoverageExceptions = []reflect.Type{
+		AbsoluteTimestamps: false,
+		ColumnLabels:       []string{},
+	}
+	printer := NewHumanReadablePrinter(&printOptions)
```
This should return the printer with default options:

```go
printer := NewHumanReadablePrinter(nil)
```

rhcarvalho commented on pkg/cmd/util/clientcmd/factory.go:
```diff
@@ -375,8 +375,14 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
-	w.Printer = func(mapping *meta.RESTMapping, noHeaders, withNamespace, wide bool, showAll bool, showLabels, absoluteTimestamps bool, columnLabels []string) (kubectl.ResourcePrinter, error) {
-		return describe.NewHumanReadablePrinter(noHeaders, withNamespace, wide, showAll, showLabels, absoluteTimestamps, columnLabels), nil
+	w.Printer = func(mapping *meta.RESTMapping, options *kubectl.PrintOptions) (kubectl.ResourcePrinter, error) {
+		if mapping != nil {
```
Maybe this needs a comment or something, it's not immediately clear what this block means.
Why do we set `options.KindName = mapping.Resource`?! And later maybe change it again.

Today we know this is about force printing resource names (or short names) when there are multiple objects types in the same output list... but what about us maintaining this code a week from now? :)

Hmm, I'm not even sure my sentence above is right, as this code doesn't seem to represent it.

What is this meant to do? :)

juanvallejo replied:
@rhcarvalho Thanks for all of your feedback!
> Why do we set options.KindName = mapping.Resource?!

Since this is the only file in the resource-printing process where we can access our own resource short forms from the same package where they are defined, I immediately assign the long resource name to the `PrintOptions` struct, just in case there is no alias for it:
```
options.KindName = mapping.Resource
```

Then, I look for our alias definition:

```
if alias, ok := resourceShortFormFor(mapping.Resource); ok {
	options.KindName = alias
}
```

If no alias is found, I simply leave the original long resource name. Then, once we jump back to `get.go` in the kubernetes package, it will see that we have already set a `KindName` from our list of resource aliases, and leave that as it is:

```
resourceName := resourcePrinter.GetResourceName() // Retrieves the name we set in the factory.go `Printer` method
if resourceName == "" {
	resourceName = mapping.Resource
}
```

rhcarvalho replied:
After I commented here, and just a few minutes ago, I had a suggestion for the similar `resourceShortFormFor` in the Kubernetes code that would simplify matters here too.
https://github.com/openshift/origin/pull/9334/files/1ff8ec94d186fe3770c06cdece2ed8d5d3d9ef99#r69644282

It's a bit strange, why does this Printer automagically sets its own resource type?
Don't we have code to set the type when printing anyway?

juanvallejo commented:
1ff8ec9 has a related upstream PR: https://github.com/kubernetes/kubernetes/pull/28509

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -397,7 +407,17 @@ func Example_printReplicationControllerWithNamespace() {
 func Example_printMultiContainersReplicationControllerWithWide() {
 	f, tf, codec := NewAPIFactory()
-	tf.Printer = kubectl.NewHumanReadablePrinter(false, false, true, false, false, false, []string{})
+	tf.Printer = kubectl.NewHumanReadablePrinter(&kctl.PrintOptions{
```
```go
tf.Printer = kubectl.NewHumanReadablePrinter(&kctl.PrintOptions{Wide: true})
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -446,7 +466,17 @@ func Example_printMultiContainersReplicationControllerWithWide() {
 func Example_printReplicationController() {
 	f, tf, codec := NewAPIFactory()
-	tf.Printer = kubectl.NewHumanReadablePrinter(false, false, false, false, false, false, []string{})
+	tf.Printer = kubectl.NewHumanReadablePrinter(&kctl.PrintOptions{
```
```go
tf.Printer = kubectl.NewHumanReadablePrinter(nil)
```

rhcarvalho replied:
Similar in other parts below.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -76,7 +76,18 @@ func NewCmdGet(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+		AbsoluteTimestamps: false,
+		ColumnLabels:       []string{},
+	}
+	p, err := f.Printer(nil, &printOptions)
```
```go
p, err := f.Printer(nil, nil)
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -310,7 +328,21 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+		if resourcePrinter, found := printer.(*kubectl.HumanReadablePrinter); found {
+			resourceName := resourcePrinter.GetResourceName()
+			if resourceName == "" {
+				resourceName = mapping.Resource
```
If I followed the logic above correctly, then `mapping` may be `nil` here and this line may panic.
I hope `f.PrinterForMapping(cmd, mapping, allNamespaces)` handles a `nil` value for `mapping`...

Before accessing some value from `mapping` we need to make sure it is not `nil`.

rhcarvalho replied:
@juanvallejo don't we need a nil check here?

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -114,6 +126,7 @@ func NewCmdGet(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *GetOptions) error {
 	selector := cmdutil.GetFlagString(cmd, "selector")
 	allNamespaces := cmdutil.GetFlagBool(cmd, "all-namespaces")
+	allKinds := cmdutil.GetFlagBool(cmd, "show-kind")
```
I think the naming here is confusing. Why is this called `allKinds` while the flag is `show-kind`?
Then later we say `printWithKinds := allKinds`. That's confusing to follow, at least for me.

It seems that we don't need both `allKinds` and `printWithKinds`.

How about:

```go
showKind := cmdutil.GetFlagBool(cmd, "show-kind")
// ...

if mustPrintWithKinds(objs, infos, sorter) {
	showKind = true
}
```

The variable name follows the flag name, and we force-set it to `true` if we must.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -323,3 +355,32 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 }
+
+// mustPrintWithKinds determines if printer output
+// should print multiple kinds of objects, even if
```
s/should/must

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -323,3 +355,32 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+
+// mustPrintWithKinds determines if printer output
+// should print multiple kinds of objects, even if
+// "all-kinds" flag is not set
```
The flag name is "show-kind".
Anyway, this function has nothing to do with the flag. It doesn't need to know that flag exist nor anything.

While it's tempting to mention it while it's hot in our minds, long term it's a bad idea. For example, imagine we rename the flag to "prefix-with-kind", or anything ("all-kind"!) -- this comment will become rotten because nobody will come look at this function when changing the flag.

It doesn't help understand what the function does, please let's get rid of it :)

rhcarvalho replied:
Instead, it would be nice to explain *why* this function will return true, under which circumstance it does so.

(And the answer is when there are objects of multiple kinds, right?).

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -323,3 +355,32 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+	var lastMap *meta.RESTMapping
+	var withKind bool
+
+	if withKind {
```
This if statement is dead code, `withKind` was just declared a line above, and it is always false at this point.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -323,3 +355,32 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+
+		// display "kind" column only if we have mixed resources
+		if lastMap != nil && mapping.Resource != lastMap.Resource {
+			withKind = true
```
Once more we can apply the idea of terminating the loop as soon as we know our answer :)

If we see two resources of different kinds, we can return immediately:

```go
		if lastMap != nil && mapping.Resource != lastMap.Resource {
			return true
		}
		lastMap = mapping
	}
	return false
}
```

Notice we don't need the `withKind` variable at all.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go:
```diff
@@ -323,3 +355,32 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+			mapping = infos[sorter.OriginalPosition(ix)].Mapping
+		} else {
+			mapping = infos[ix].Mapping
+		}
```
I feel this is some not very trivial logic. A similar loop was part of `RunGet` before, so it probably works and very likely has no tests.

Now that we have this `mustPrintWithKinds` function, it wouldn't be hard to write a few unit tests that exercise it, so that we also make sure this logic here is reasonable.

We could have a few objects all of the same kind, and `mustPrintWithKinds` should return false, and another test case where we have two or more different kinds, and `mustPrintWithKinds` returns true. And another test where sorter != nil so we exercise the two branches of the if.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go:
```diff
@@ -1053,7 +1053,19 @@ func (f *Factory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMappin
+			NoHeaders:          GetFlagBool(cmd, "no-headers"),
+			WithNamespace:      withNamespace,
+			WithKind:           false,
+			KindName:           "",
```
You can omit `WithKind` and `KindName`, since then are always the zero value.

Isn't it valid to use the "show-kind" flag here?

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -226,7 +231,17 @@ func ErrorPrintHandler(obj *TestPrintType, w io.Writer, options PrintOptions) er
 func TestCustomTypePrinting(t *testing.T) {
 	columns := []string{"Data"}
-	printer := NewHumanReadablePrinter(false, false, false, false, false, false, []string{})
+	printer := NewHumanReadablePrinter(&PrintOptions{
```
Same here about using the zero values, in all occurrences in this file :)

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -670,7 +773,7 @@ func TestPrintHunmanReadableIngressWithColumnLabels(t *testing.T) {
 	}
 	buff := bytes.Buffer{}
-	printIngress(&ingress, &buff, PrintOptions{false, false, false, false, false, false, []string{"app_name"}})
+	printIngress(&ingress, &buff, PrintOptions{false, false, false, false, false, false, false, "", []string{"app_name"}})
```
I'd suggest:

```go
printIngress(&ingress, &buff, PrintOptions{ColumnLabels: []string{"app_name"}})
```

Most times it's a bad idea to initialize a big struct like that with positional fields instead of explicit field names.
With so many boolean fields, if somebody would reorder or add more fields (like we're doing) code like that would compile, while potentially changing the semantics (a field that used to be initialized to false accidentally becoming true and so on).

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -792,7 +895,7 @@ func TestPrintHumanReadableService(t *testing.T) {
 	for _, svc := range tests {
 		buff := bytes.Buffer{}
-		printService(&svc, &buff, PrintOptions{false, false, false, false, false, false, []string{}})
+		printService(&svc, &buff, PrintOptions{false, false, false, false, false, false, false, "", []string{}})
```
Using the zero value of PrintOptions:

```go
printService(&svc, &buff, PrintOptions{})
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -971,10 +1074,22 @@ func TestPrintHumanReadableWithNamespace(t *testing.T) {
+		AbsoluteTimestamps: false,
+		ColumnLabels:       []string{},
+	}
+
```
We can get rid of this and just use `nil` in all spots below to get a printer with default options.

juanvallejo commented:
@rhcarvalho Thanks again for all of the feedback, I have addressed your suggestions in https://github.com/openshift/origin/pull/9334/commits/03981d1aa40a6ea766afb0d731bdb867e358cd4f

juanvallejo commented:
@rhcarvalho 
>It's a bit strange, why does this Printer automagically sets its own resource type?
Don't we have code to set the type when printing anyway?

It is not possible to access our own aliases when printing, as including the `clientcmd` package from the `describe` package causes a circular dependency. Setting it here:

```
w.Printer = func(mapping *meta.RESTMapping, options *kubectl.PrintOptions) (kubectl.ResourcePrinter, error) {
	if mapping != nil {
		options.KindName = mapping.Resource
		if alias, ok := resourceShortFormFor(mapping.Resource); ok {
			options.KindName = alias
		}
	}
	return describe.NewHumanReadablePrinter(options), nil
}
```

is what I did to prevent passing our resource alias map as an argument to the HumanResourcePrinter as I was doing originally. Also, there is no way to access the actual Resource Kind from the resource printer for a specific resource: https://github.com/kubernetes/kubernetes/pull/27379#discussion_r68997841

rhcarvalho commented:
> there is no way to access the actual Resource Kind from the resource printer for a specific resource: kubernetes/kubernetes#27379 (comment)

Yeah that's sad. Funny that I came across the same problem a while ago when working on https://github.com/feedhenry/openshift-template-tool.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/edit.go:
```diff
@@ -88,7 +88,17 @@ func NewCmdEdit(f *cmdutil.Factory, out, errOut io.Writer) *cobra.Command {
 	// retrieve a list of handled resources from printer as valid args
 	validArgs, argAliases := []string{}, []string{}
-	p, err := f.Printer(nil, false, false, false, false, false, false, []string{})
+	p, err := f.Printer(nil, &kubectl.PrintOptions{
```
One more for the zero value gang :)

```go
p, err := f.Printer(nil, nil)
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/label.go:
```diff
@@ -74,7 +74,17 @@ func NewCmdLabel(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	// retrieve a list of handled resources from printer as valid args
 	validArgs, argAliases := []string{}, []string{}
-	p, err := f.Printer(nil, false, false, false, false, false, false, []string{})
+	p, err := f.Printer(nil, &kubectl.PrintOptions{
```
```go
p, err := f.Printer(nil, nil)
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/patch.go:
```diff
@@ -65,7 +65,17 @@ func NewCmdPatch(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	// retrieve a list of handled resources from printer as valid args
 	validArgs, argAliases := []string{}, []string{}
-	p, err := f.Printer(nil, false, false, false, false, false, false, []string{})
+	p, err := f.Printer(nil, &kubectl.PrintOptions{
```
```go
p, err := f.Printer(nil, nil)
```

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -320,10 +320,12 @@ type handlerEntry struct {
 	ShowAll            bool
 	ShowLabels         bool
 	AbsoluteTimestamps bool
+	KindName           string
```
`WithKind` feels like the other controls / print options, it tells the printer whether to show or not kind prefixes next to object names.
But `KindName` is really strange. The more I think about it, it seems to be just some resource-specific value hidden into the printer.

Why does a printer need to know about a kind name?! Isn't the printer generic and able to print resources of different kinds?!

juanvallejo replied:
> Why does a printer need to know about a kind name?! Isn't the printer generic and able to print resources of different kinds?!

I agree with this. At least for kube resources, I pass the resource alias (full name if no alias) to a printer method here: https://github.com/openshift/origin/pull/9334/commits/259f9890c88b2dafaa7d95cd9ed1ae9c343ad384#diff-85d15d1be2a3707c4cc1b4e658dff84aR363

Perhaps making a separate method, maybe something like `SetResourceAlias`, would work better. That way, I can call it from `get.go` for kube resources, and from our `factory.go` Printer method here: https://github.com/openshift/origin/pull/9334/commits/8bf8c5532adbd09ed337fc108146c7e5cc6568e8#diff-0073015bc0e88b3269344b5a87d471dbR380 for our resources

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -707,6 +727,11 @@ func printReplicationController(controller *api.ReplicationController, w io.Writ
 	name := controller.Name
 	namespace := controller.Namespace
 	containers := controller.Spec.Template.Spec.Containers
+	kind := options.KindName
```
This adds to my comment https://github.com/openshift/origin/pull/9334/files#r69782642.

We are printing a Replication Controller. The kind is "replicationcontroller" or "rc". The "kind name" shouldn't be something that comes from the printer options... There's a mismatch in terms of who should know what.

juanvallejo replied:
I will change `options.KindName` to `options.Kind` as it will better reflect its value

juanvallejo commented:
```
> @@ -310,7 +328,21 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
>  			}
>  			lastMapping = mapping
>  		}
> -		if _, found := printer.(*kubectl.HumanReadablePrinter); found {
> +		if resourcePrinter, found := printer.(*kubectl.HumanReadablePrinter); found {
> +			resourceName := resourcePrinter.GetResourceName()
> +			if resourceName == "" {
> +				resourceName = mapping.Resource
@juanvallejo don't we need a nil check here?
```

@rhcarvalho Done!

juanvallejo commented:
flake? re[test]

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -338,23 +340,31 @@ type HumanReadablePrinter struct {
 
 // NewHumanReadablePrinter creates a HumanReadablePrinter.
-func NewHumanReadablePrinter(noHeaders, withNamespace bool, wide bool, showAll bool, showLabels bool, absoluteTimestamps bool, columnLabels []string) *HumanReadablePrinter {
+func NewHumanReadablePrinter(printOptions *PrintOptions) *HumanReadablePrinter {
```
If all we do with `printOptions` is an indirection, `options : *printOptions`, why do we make the argument be a pointer type?

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -241,9 +246,26 @@ func TestCustomTypePrinting(t *testing.T) {
+func TestCustomTypePrintingWithKind(t *testing.T) {
+	columns := []string{"Data"}
+	printer := NewHumanReadablePrinter(&PrintOptions{WithKind: true, Kind: "test"})
+	printer.Handler(columns, PrintCustomType)
```
I got the feeling this test is really just testing `PrintCustomType`, which is only used in tests...

This is related to one of the first comments I've made, expressing my concern for the implementation requiring code in every print function for every type. If we forget to add that logic to one type, we'll be broken and there are no tests to capture it.

rhcarvalho commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -241,9 +246,26 @@ func TestCustomTypePrinting(t *testing.T) {
+		t.Errorf("The data was not printed as expected. Expected:\n%s\nGot:\n%s", expectedOutput, buffer.String())
+	}
+}
+
```
We're missing a test case in which the printer is created with WithKind:false, but when we try to print objects of multiple types it should force WithKind to be true.

rhcarvalho commented on pkg/cmd/util/clientcmd/factory.go:
```diff
@@ -375,8 +375,16 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
-	w.Printer = func(mapping *meta.RESTMapping, noHeaders, withNamespace, wide bool, showAll bool, showLabels, absoluteTimestamps bool, columnLabels []string) (kubectl.ResourcePrinter, error) {
-		return describe.NewHumanReadablePrinter(noHeaders, withNamespace, wide, showAll, showLabels, absoluteTimestamps, columnLabels), nil
+	// Saves current resource name (or alias if any) in PrintOptions. Once saved, it will not be overwritten by the
+	// kubernetes resource alias look-up, as it will notice a non-empty value in `options.KindName`
```
The comment is outdated, s/KindName/Kind

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -161,6 +161,13 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
+// Receives a resource kind and its name, returns a formatted
+// string with the "resource_kind/resource_name"
+func formatResourceName(kind, name string) string {
+	rname := kind + "/" + name
```
Suggestion:

```go
func TestFormatResourceName(t *testing.T) {
	tests := []struct {
		kind, name string
		want       string
	}{
		{"", "", ""},
		{"", "name", "name"},
		{"kind", "", "kind/"}, // should not happen in practice
		{"kind", "name", "kind/name"},
	}
	for _, tt := range tests {
		if got := formatResourceName(tt.kind, tt.name); got != tt.want {
			t.Errorf("formatResourceName(%q, %q) = %q, want %q", tt.kind, tt.name, got, tt.want)
		}
	}
}
```

Now, if you make `formatResourceName` pass those tests, you don't need to write a lot of ifs in every `print*` function, you could always say:

```go
name := formatResouceName(opts.Kind, v.Name) // where v is t, build, bc, etc, depending on the arg.
// No need to store opts.Kind in a variable, it is ever only used above.
```

WDYT?

Would be nice to follow the same pattern in the Kubernetes code. There we're not using anything like `formatResouceName`, but repeating the ifs and concatenating the strings in every place.

juanvallejo replied:
@rhcarvalho Thanks again for the feedback! I do think having something like the above suggestion would be cleaner. I'll go ahead and put it together.

juanvallejo commented:
@fabianofranz Could you take a look please?

rhcarvalho commented on contrib/completions/bash/oadm:
```diff
@@ -2416,7 +2416,6 @@ _oadm_taint()
     must_have_one_noun+=("appliedclusterresourcequota")
     must_have_one_noun+=("build")
     must_have_one_noun+=("buildconfig")
-    must_have_one_noun+=("cluster")
```
This is strange... how is it related to this PR?
@juanvallejo maybe you need a rebase?

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -161,6 +161,21 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
+		slash = ""
+	}
+
+	return (kind + slash + name)
```
The parentheses here are unnecessary, and not common use in Go (see for example the `if` and `for` statements).

rhcarvalho commented:
@juanvallejo could you please add a `TestFormatResourceName` to origin code as well?
We have the function in both places, good to have the accompanying tests :)

rhcarvalho commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -161,6 +161,21 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
 }
 
+// Receives a resource kind and its name, returns a formatted
+// string with the "resource_kind/resource_name"
```
The comment is outdated, and should follow the convention of `formatResourceName (...).`

rhcarvalho commented on vendor/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -844,39 +868,8 @@ func printReplicaSetList(list *extensions.ReplicaSetList, w io.Writer, options P
 	return nil
 }
 
-func printCluster(c *federation.Cluster, w io.Writer, options PrintOptions) error {
```
This removal along with other changes to "cluster", looks suspicious.
@juanvallejo could you please confirm this is intentional? Why is it being removed?

juanvallejo replied:
@rhcarvalho Hm, I had assumed these had been removed as part of the rebase update, but they still [seem to be there in master](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/resource_printer.go#L35). Thanks for catching this, I will take a look right now

rhcarvalho replied:
> @rhcarvalho Hm, I had assumed these had been removed as part of the
rebase update, but they still seem to be there in master. Thanks for
catching this, I will take a look right now

I imagined it could be coming from there. In any case it should not be part
of your changes, it should have been handled by some other PR, since we
always check for those generated files being up to date :-)

rhcarvalho commented on vendor/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -32,7 +32,6 @@ import (
 
 	"github.com/ghodss/yaml"
 	"github.com/golang/glog"
-	"k8s.io/kubernetes/federation/apis/federation"
```
Suspicious.

juanvallejo replied:
@rhcarvalho This was most likely taken out after rebase with new version of kube

rhcarvalho commented on vendor/k8s.io/kubernetes/pkg/kubectl/resource_printer_test.go:
```diff
@@ -241,9 +263,44 @@ func TestCustomTypePrinting(t *testing.T) {
+func TestCustomTypePrintingWithForcedKind(t *testing.T) {
+	columns := []string{"Data"}
+	printer := NewHumanReadablePrinter(&PrintOptions{WithKind: false})
+	printer.EnsurePrintWithKind("test")
```
We should not call this method in the test.
What we should be testing is that when the input contains multiple types, then the printer will automatically include the kinds.

If the test calls `EnsurePrintWithKind` it is cheating.

juanvallejo replied:
Ah, I am not sure if this can be truly approached from here, as it is determined in [kubernetes/pkg/kubectl/cmd/get.go](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/cmd/get.go#L307) whether or not kind will be included. Another approach to testing this could be to make a test in `hack/test-cmd.sh` and expect an output of multiple known resources to be formatted as `kind/name`

juanvallejo commented:
@rhcarvalho 
Ah, I am not sure if this can be truly approached from here, as it is determined in [kubernetes/pkg/kubectl/cmd/get.go](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/cmd/get.go#L307) whether or not kind will be included. Another approach to testing this could be to make a [test in `hack/test-cmd.sh` and expect an output of multiple known resources to be formatted as `kind/name`](https://github.com/juanvallejo/origin/blob/240c76927d32197fcfa8fdb03f580cd76d33c56a/test/cmd/basicresources.sh#L266-272)

juanvallejo commented:
@rhcarvalho 
> could you please add a TestFormatResourceName to origin code as well?
We have the function in both places, good to have the accompanying tests :)

Ah, did not see this earlier. Done! https://github.com/openshift/origin/commit/240c76927d32197fcfa8fdb03f580cd76d33c56a

rhcarvalho commented on test/cmd/basicresources.sh:
```diff
@@ -263,6 +263,14 @@ os::cmd::expect_success_and_text "oc get route route-with-set-port --template='{
+os::cmd::expect_success 'oc new-app node'
+os::cmd::expect_success_and_text 'oc get all' 'is/test1'
+echo "resource printer: ok"
+os::test::junit::declare_suite_end
```
:+1: @juanvallejo this works for me for validating that the kind gets automatically added.

Of course there could still be other kinds that will print without the prefix, but that's a limitation we seen to be accepting. For example, printing a new kind will require adding some code to the printer of that kind.

As for this test here and now, @juanvallejo do you think it would make sense to check that `oc get is` does not include the `is/` prefix, while `oc get all` does?

juanvallejo replied:
@rhcarvalho 
> do you think it would make sense to check that oc get is does not include the is/ prefix, while oc get all does?

Sure, I'll go ahead and add a test case for this

rhcarvalho commented:
@juanvallejo thanks for being patient with all of my comments :)
Add that test case and LGTM.

@fabianofranz how has been your policy for commits in a PR? @bparees is normally in favor of a squash before merge.

fabianofranz commented:
@rhcarvalho thanks for your review, yes, @juanvallejo squash into 3 commits (both UPSTREAM ones and one Origin commit) and let's get this in. Let me know when you are done with the test case and squashing.

juanvallejo commented:
@fabianofranz 
> squash into 3 commits (both UPSTREAM ones and one Origin commit) and let's get this in

Done!

fabianofranz commented:
[merge]

@juanvallejo please note that if the upstream changes were significant in comparison to what was initially merged there, you may need to open a new PR upstream with the changes.

juanvallejo commented:
@fabianofranz Thanks! I have done so here: https://github.com/kubernetes/kubernetes/pull/28509 . I will be updating it today to reflect the rest of the changes made in this PR.

fabianofranz commented:
@juanvallejo well done! Make sure to put the related BZ status to ON_QA so that they can test it. ;)
